storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java

```diff
@@ -462,22 +462,32 @@ public Optional<ProducerStateEntry> lastEntry(long producerId) {
     }
 
     /**
-     * Take a snapshot at the current end offset if one does not already exist.
+     * Take a snapshot at the current end offset if one does not already exist with syncing the change to the device
      */
     public void takeSnapshot() throws IOException {
+        takeSnapshot(true);
+    }
+
+    /**
+     * Take a snapshot at the current end offset if one does not already exist, then return the snapshot file if taken.
+     */
+    public Optional<File> takeSnapshot(boolean sync) throws IOException {
```

Comment:
`ProducerStateManager.truncateFullyAndReloadSnapshots` removes all snapshot files and then calls `loadSnapshots()`, which should return empty. I am wondering what happens if we have an pending async snapshot flush and the flush is called after the underlying file is deleted because of `ProducerStateManager.truncateFullyAndReloadSnapshots`. Will that cause the file to be recreated or will it get an `IOException`? The former will be bad since the content won't be correct. For the latter, it would be useful to distinguish that from a real disk IO error to avoid unnecessarily crash the broker.

Reply:
Thanks, that's a good point. I overlooked snapshot files would be cleaned up upon receiving `OffsetMovedToRemoteStorage`.

In this case, if async flush is performed against non-existent file, it would throw IOException so we should catch it and ignore if it's NoSuchFileException.
(Since file creation is still done in original thread. Only fsync is moved to async thread)

I'll fix that